### PR TITLE
Migrate to PostgreSQL

### LINKED ISSUE #83: POST requests returns incorrect ancestors

### LINKED ISSUE #171: Lock route while attaching or detaching point

### LINKED ISSUE #238: Migrate to PostgreSQL and use ltree

### PR DESCRIPTION
**Pull request by kasper-i**

### CODE REVIEW

File: backend/model/resource.go [hunk 1]
```diff
@@ -314,13 +318,13 @@ func (sess Session) updateCountersForResource(resourceID string, delta Counters)
 		return nil
 	}
 
-	var params []string = make([]string, 0)
+	var param string = "counters"
 
 	for counterType, count := range difference {
-		params = append(params, fmt.Sprintf("'$.%s', (COALESCE(JSON_EXTRACT(counters, '$.%s'), 0) + %d) DIV 1", counterType, counterType, count))
+		param = fmt.Sprintf("jsonb_set(%s::jsonb, '{%s}', DIV((COALESCE((counters->>'%s')::int, 0) + %d), 1)::text::jsonb, true)", param, counterType, counterType, count)
```
Comment (kasper-i) [inline backend/model/resource.go:382]: Can we omit COALESCE?

File: web/src/config.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-  "API_URL": "https://api.bultdatabasen.se"
+  "API_URL": "http://localhost:8080"
```
Comment (kasper-i) [inline web/src/config.json]:
```suggestion
  "API_URL": "https://api.bultdatabasen.se"
```

File: model/postgres/post.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+CREATE OR REPLACE FUNCTION
+  populate_path(ID varchar)
+RETURNS
+  void
+AS $$
+DECLARE
+
+BEGIN
+  RAISE NOTICE 'A simple text';
+END $$ LANGUAGE plpgsql;
```
Comment (HerrNaN) [inline model/postgres/post.sql]: What purpose does this fill?

Comment (kasper-i) [reply model/postgres/post.sql]: Nothing at the moment. Will make this a procedure that builds up the new ltree paths.

Comment (kasper-i) [reply model/postgres/post.sql]: Take a look now! :wink:

File: backend/model/resource.go [hunk 4]
```diff
@@ -118,96 +115,108 @@ func (sess Session) GetResource(resourceID string) (*Resource, error) {
...
-	var resource Resource
-
-	if err := sess.DB.Raw(`SELECT * FROM resource WHERE id = ? FOR UPDATE`, resourceID).
-		Scan(&resource).Error; err != nil {
-		return nil, err
-	}
-
-	if resource.ID == "" {
-		return nil, gorm.ErrRecordNotFound
+func (sess Session) getSubtreeLock(resourceID string) error {
```
Comment (kasper-i) [inline backend/model/resource.go]: Not locking leaf nodes.

Comment (kasper-i) [reply backend/model/resource.go]: Fixed!

File: backend/model/bolt.go [hunk 5]
```diff
@@ -37,26 +36,22 @@ func (bolt *Bolt) UpdateCounters() {
...
 
-	resourceType := "bolt"
-
-	query := fmt.Sprintf(`%s
-	SELECT
-		%s.*,
-		resource.parent_id,
+	query := fmt.Sprintf(`%s SELECT
+		bolt.*,
+		resource.leaf_of AS parent_id,
```
Comment (kasper-i) [inline backend/model/bolt.go]:
```suggestion
```

File: backend/model/bolt.go [hunk 6]
```diff
@@ -65,12 +60,12 @@ func (sess Session) GetBolts(resourceID string) ([]Bolt, error) {
...
 }
 
-func (sess Session) GetBolt(resourceID string) (*Bolt, error) {
+func (sess Session) GetBolt(resourceID uuid.UUID) (*Bolt, error) {
 	var bolt Bolt
 
 	if err := sess.DB.Raw(`SELECT
 			bolt.*,
-			resource.parent_id,
+			resource.leaf_of AS parent_id,
```
Comment (kasper-i) [inline backend/model/bolt.go]:
```suggestion
```

File: backend/model/db.go [hunk 7]
```diff
@@ -23,21 +23,25 @@ func init() {
...
 
-	dsn := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s?charset=utf8mb4&parseTime=True&loc=Local", username, password, host, port, database)
+	dsn := fmt.Sprintf(
+		"host=%s user=%s password=%s dbname=%s port=%s sslmode=disable search_path=%s TimeZone=Europe/Stockholm",
+		host, username, password, database, port, schema)
 
-	DB, err = gorm.Open(mysql.Open(dsn), &gorm.Config{
-		Logger: logger.Default.LogMode(logger.Warn),
+	DB, err = gorm.Open(postgres.Open(dsn), &gorm.Config{
+		Logger: logger.Default.LogMode(logger.Info),
```
Comment (kasper-i) [inline backend/model/db.go]:
```suggestion
		Logger: logger.Default.LogMode(logger.Warn),
```

File: backend/model/route.go [hunk 8]
```diff
@@ -89,36 +90,45 @@ func (sess Session) CreateRoute(route *Route, parentResourceID string) error {
...
 		updatedRoute.ID = original.ID
-		updatedRoute.ParentID = original.ParentID
 		updatedRoute.Counters = original.Counters
 		updatedRoute.UpdateCounters()
 
 		countersDifference := updatedRoute.Counters.Substract(original.Counters)
 
-		if err := sess.touchResource(routeID); err != nil {
+		if updatedRoute.Name != original.Name {
+			sess.RenameResource(routeID, updatedRoute.Name)
```
Comment (kasper-i) [inline backend/model/route.go]: Add missing error check.

File: web/src/config.json (hunk 2 above)
Comment (kasper-i) [inline web/src/config.json]:
```suggestion
  "API_URL": "https://api.bultdatabasen.se"
```

File: backend/model/common.go [hunk 9]
```diff
@@ -24,101 +24,118 @@ func (sess Session) Transaction(fn func(sess Session) error) error {
...
-	if resource.ParentID == nil {
-		return utils.ErrOrphanedResource
+	if err := sess.DB.Raw(`SELECT * FROM resource WHERE id = ? FOR UPDATE`, resourceID).Scan(&resource).Error; err != nil {
+		return nil, err
 	}
 
-	if !sess.checkParentAllowed(resource, *resource.ParentID) {
-		return utils.ErrHierarchyStructureViolation
-	}
+	return &resource, nil
```
Comment (kasper-i) [inline backend/model/common.go:42]: Missing not found check?

File: backend/model/common.go [hunk 10]
```diff
@@ -137,7 +154,7 @@ func (sess Session) deleteResource(resourceID string) error {
 	return nil
 }
 
-func (sess Session) checkParentAllowed(resource Resource, parentID string) bool {
+func (sess Session) checkParentAllowed(resource Resource, parentID uuid.UUID) bool {
```
Comment (kasper-i) [inline backend/model/common.go:161]: Check leafs as well?

File: backend/model/resource.go [hunk 11]
```diff
@@ -118,96 +135,121 @@ func (sess Session) GetResource(resourceID string) (*Resource, error) {
...
 	var err error
+	var oldParentID uuid.UUID
 
 	return sess.Transaction(func(sess Session) error {
-		if resource, err = sess.getResourceWithLock(resourceID); err != nil {
+		if err := sess.getSubtreeLock(resourceID); err != nil {
+			return err
+		}
+
+		if resource, err = sess.GetResource(resourceID); err != nil {
```
Comment (kasper-i) [inline backend/model/resource.go]: Take lock.

File: backend/model/resource.go [hunk 12]
```diff
@@ -118,96 +135,121 @@ func (sess Session) GetResource(resourceID string) (*Resource, error) {
...
 
-func (sess Session) getResourceWithLock(resourceID string) (*Resource, error) {
-	var resource Resource
-
-	if err := sess.DB.Raw(`SELECT * FROM resource WHERE id = ? FOR UPDATE`, resourceID).
-		Scan(&resource).Error; err != nil {
-		return nil, err
-	}
+func (sess Session) RenameResource(resourceID uuid.UUID, name string) error {
+	return sess.DB.Exec(`UPDATE resource SET name = ? WHERE id = ?`, name, resourceID).Error
```
Comment (kasper-i) [inline backend/model/resource.go]: Set modified timestamp.

File: backend/model/point.go [hunk 13]
```diff
@@ -15,42 +17,34 @@ type Point struct {
...
-				WHERE id IN ?
-			UNION
-				SELECT id, foster_parent_id AS parent_id, TRUE as foster_care
-				FROM foster_care
-				WHERE id IN ?
-		) AS child
-		INNER JOIN resource parent ON child.parent_id = parent.id`, pointIDs, pointIDs).Scan(&parents).Error
+	err := sess.DB.Raw(`SELECT id, name, type, tree.resource_id as child_id
+		FROM tree
+		INNER JOIN resource parent ON REPLACE(subpath(tree.path, -2, 1)::text, '_', '-')::uuid = parent.id
```
Comment (kasper-i) [inline backend/model/point.go:39]: Use lquery here instead.

File: model/postgres/post.sql [hunk 14]
```diff
@@ -0,0 +1,244 @@
+ALTER SCHEMA old RENAME TO bultdatabasen;
```
Comment (kasper-i) [inline model/postgres/post.sql:1]: Remove MySQL schema.

File: model/postgres/bultdatabasen.sql [hunk 15]
```diff
@@ -0,0 +1,1114 @@
+--
+-- PostgreSQL database dump
```
Comment (kasper-i) [inline model/postgres/bultdatabasen.sql:2]: Re-dump latest version.

File: model/postgres/post.sql [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+    END LOOP;
+END $$ LANGUAGE plpgsql;
+
+SELECT populate_path(id) FROM resource WHERE type <> 'root' AND depth < 600 AND parent_id IS NOT NULL;
+
+ALTER TABLE resource DROP COLUMN depth;
+UPDATE resource SET parent_id = NULL WHERE type IN ('area', 'crag', 'point', 'root', 'route', 'sector');
+DROP TABLE foster_care;
+
+-- mode resource types into resource_type enum type
```
Comment (kasper-i) [inline model/postgres/post.sql]:
```suggestion
-- move resource types into resource_type enum type
```

File: model/postgres/post.sql [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+ALTER TABLE user_role ALTER COLUMN role TYPE role USING role::role;
+ALTER TABLE team_role ALTER COLUMN role TYPE role USING role::role;
+
+-- rename to leaf_of to catch every query that needs to be rewritten
+
+ALTER TABLE resource RENAME COLUMN parent_id TO leaf_of;
+
+-- enhance trash bin
+
+ALTER TABLE trash RENAME COLUMN orig_parent_id TO orig_leaf_of;
```
Comment (kasper-i) [inline model/postgres/post.sql:242]: Trash bin is not migrated.

File: model/postgres/post.sql [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+ALTER TABLE trash ALTER COLUMN orig_leaf_of DROP NOT NULL;
+ALTER TABLE trash ADD COLUMN 
+ALTER TABLE trash ADD orig_path ltree;
```
Comment (kasper-i) [inline model/postgres/post.sql]:
```suggestion
ALTER TABLE trash ADD COLUMN orig_path ltree;
```

**Pull request merged**